typing/ctype.ml

```diff
@@ -4916,9 +4954,30 @@ let immediacy env typ =
             | _ -> false)
           row.row_fields
       then
-        Type_immediacy.Unknown
+        Type_immediacy.coerce Unknown ~as_:imm
       else
-        Type_immediacy.Always
-  | _ -> Type_immediacy.Unknown
-
-let maybe_pointer_type env typ = not (is_immediate (immediacy env typ))
+        Type_immediacy.coerce Always ~as_:imm
+  | _ -> Type_immediacy.coerce Unknown ~as_:imm
+
+let check_decl_immediate env decl imm =
+  match decl with
+  | { type_unboxed = { unboxed = true; _ };
+      type_kind = ( Type_variant [{cd_args = Cstr_tuple [arg]; _}]
+                  | Type_variant [{cd_args = Cstr_record [{ld_type=arg; _}]; _}]
+                  | Type_record ([{ld_type=arg; _}], _)); _ } ->
+     check_type_immediate env arg imm
+  | { type_kind; type_manifest = None; _ } ->
+     Type_immediacy.coerce (kind_immediacy type_kind) ~as_:imm
+  | { type_kind; type_manifest = Some ty; _ } ->
+     (* Check the kind first, in case of missing cmis *)
+     match Type_immediacy.coerce (kind_immediacy type_kind) ~as_:imm with
+     | Ok () -> Ok ()
+     | _ -> check_type_immediate env ty imm
```

Comment:
After making the other suggested changes, I think you should be able to share more code for this function.